overlord/devicestate/handlers_systems.go

```diff
@@ -157,10 +159,40 @@ func (m *DeviceManager) doCreateRecoverySystem(t *state.Task, _ *tomb.Tomb) (err
 			return nil, false, err
 		}
 		logger.Debugf("requested info for not yet installed snap %q", name)
-		// TODO: handle remodel case in which snap may not be installed
-		// yet, and thus we need to pull info from snapsup of relevant
-		// download tasks
-		return nil, false, fmt.Errorf("not implemented")
+
+		if !isRemodel {
+			// when not in remodel, a recovery system can only be
+			// created from snaps that are already installed
+			return nil, false, nil
+		}
+
+		// in a remodel scenario, the snaps may need to be fetched, and
+		// thus we can pull the relevant information from the tasks
+		// carrying snap-setup
+
+		for _, tskID := range setup.SnapSetupTasks {
+			taskWithSnapSetup := st.Task(tskID)
+			snapsup, err := snapstate.TaskSnapSetup(taskWithSnapSetup)
+			if err != nil {
+				return nil, false, err
+			}
+			if snapsup.SnapName() != name {
+				continue
+			}
+			// by the time this task runs, the file has already been
+			// downloaded
```

Comment:
and validated  (we split the phases after validate-snap I think),  which means the open/read below are not risky